### PR TITLE
Testability transformation for Jackson

### PR DESCRIPTION
**Pull request by seran**

### CODE REVIEW

**Comment by seran** [comment]
Not sure about the implementation in `JacksonObjectMapperClassReplacement.java` at line 73. It works, but not sure whether it's the correct approach.

**Comment by arcuri82** [comment]
@seran remind me tomorrow, we can look at it together

File: client-java/instrumentation/pom.xml [hunk 1]
```diff
@@ -97,6 +97,14 @@
             <artifactId>jackson-annotations</artifactId>
             <scope>test</scope>
         </dependency>
+        <dependency>
```
Comment (ZhangMan1126) [inline client-java/instrumentation/pom.xml:100]: check it if these two dependencies are only applied for `test` scope.
if not, you need to add the two dependencies with the plugin
`<groupId>org.apache.maven.plugins</groupId>
 <artifactId>maven-shade-plugin</artifactId>`

see below of this `pom.xml`

File: e2e-tests/spring-rest-openapi-v3/src/test/kotlin/org/evomaster/e2etests/spring/openapi/v3/jackson/JacksonEMTest.kt [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+    companion object {
+        @BeforeAll
+        @JvmStatic
+        fun init() {
+            initClass(JacksonController())
+        }
+    }
+
+    @Test
+    fun testGenericReadValue() {
```
Comment (ZhangMan1126) [inline e2e-tests/spring-rest-openapi-v3/src/test/kotlin/org/evomaster/e2etests/spring/openapi/v3/jackson/JacksonEMTest.kt:21]: might try to split `testGenericReadValue` and `testTypeReadValue` into two test suites (ie, two test files), see whether it still fails inGA

**Review by ZhangMan1126 — CHANGES_REQUESTED** [review]

File: e2e-tests/spring-rest-openapi-v3/src/main/kotlin/com/foo/rest/examples/spring/openapi/v3/jackson/JacksonRest.kt [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+            return ResponseEntity.badRequest().body("Failed")
+        }
+        return if (x > 0) ResponseEntity.ok().body("Working")
+        else ResponseEntity.badRequest().body("Failed")
+    }
+
+    @PostMapping(path = ["/type"], consumes = [MediaType.APPLICATION_JSON_VALUE, MediaType.APPLICATION_FORM_URLENCODED_VALUE])
+    fun typeReference(request: HttpServletRequest): ResponseEntity<String> {
+
+        val json = IOUtils.toString(request.inputStream, StandardCharsets.UTF_8)
```
Comment (arcuri82) [inline e2e-tests/spring-rest-openapi-v3/src/main/kotlin/com/foo/rest/examples/spring/openapi/v3/jackson/JacksonRest.kt:36]: aren;t these 2 endpoints having exactly the same implementation?

File: client-java/instrumentation/src/main/java/org/evomaster/client/java/instrumentation/coverage/methodreplacement/thirdpartyclasses/JacksonObjectMapperClassReplacement.java [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+    @Override
+    protected String getNameOfThirdPartyTargetClass() {
+        return "com.fasterxml.jackson.databind.ObjectMapper";
+    }
+
+    @Replacement(replacingStatic = false,
+            type = ReplacementType.TRACKER,
+            id = "Jackson_ObjectMapper_readValue_InputStream_class",
+            usageFilter = UsageFilter.ONLY_SUT,
+            category = ReplacementCategory.NET)
```
Comment (arcuri82) [inline client-java/instrumentation/src/main/java/org/evomaster/client/java/instrumentation/coverage/methodreplacement/thirdpartyclasses/JacksonObjectMapperClassReplacement.java:30]: should be `EXT_0`

File: client-java/instrumentation/src/main/java/org/evomaster/client/java/instrumentation/coverage/methodreplacement/thirdpartyclasses/JacksonObjectMapperClassReplacement.java [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+        } catch (InvocationTargetException e){
+            throw (RuntimeException) e.getCause();
+        }
+    }
+
+    @Replacement(replacingStatic = false,
+            type = ReplacementType.TRACKER,
+            id = "Jackson_ObjectMapper_readValue_Generic_class",
+            usageFilter = UsageFilter.ONLY_SUT,
+            category = ReplacementCategory.BASE)
```
Comment (arcuri82) [inline client-java/instrumentation/src/main/java/org/evomaster/client/java/instrumentation/coverage/methodreplacement/thirdpartyclasses/JacksonObjectMapperClassReplacement.java:56]: should be `EXT_0`

**Review by arcuri82 — CHANGES_REQUESTED** [review]

File: e2e-tests/spring-rest-openapi-v3/src/main/kotlin/com/foo/rest/examples/spring/openapi/v3/jackson/JacksonRest.kt (hunk 3 above)
Comment (seran) [reply e2e-tests/spring-rest-openapi-v3/src/main/kotlin/com/foo/rest/examples/spring/openapi/v3/jackson/JacksonRest.kt:36]: Ah it's a mistake on the last commit, I'll correct.

**Pull request merged**